### PR TITLE
write functions  convert record to JSON incorrect when enum is involved

### LINKED ISSUE #8252: writef JSON format prints enum values incorrectly.
### Summary of Problem
When printing enum values from writef (using the `%jt` format string), the enum value string is printed without quotation marks, which is invalid JSON.
[This can be seen using an online JSON validator](http://jsoneditoronline.org/?id=4f35e24d601868831f7ab88fa5794920)

### Steps to Reproduce

**Source Code:**

```chapel
enum EnumType { Type1, Type2, Type3 };
record RecType {
  const s : string;
  const e : EnumType;
}

writef( "%jt\n", new RecType( "example_string", EnumType.Type1 )  );
```

which produces the output
```
{"s":"example_string", "e":Type1}
```

[See this TIO instance](https://tio.run/##PY5BCsIwEEX3c4pPQbBQBHVn1l5AXAqlxFEr7TQkkVakZ4@xIW7mwf8zj9GPxnAXAsurxzGO89swPvhhWy3YJewxK7KsB3vFiXXaI0AP4jwcDnDetnJX/4xjlpWKZqLRtp5vaxSrp79IUUF4zK6Y8tT0puM6eWKdjzfLNyiBUhGF8AU)

### Configuration Information

- Output of `chpl --version`: 
```
> chpl --version
chpl Version 1.16.0.e43acc7
Copyright (c) 2004-2017, Cray Inc.  (See LICENSE file for more details)

```
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
```
> $CHPL_HOME/util/printchplenv --anonymize
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none *
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_MAKE: make
CHPL_ATOMICS: intrinsics
CHPL_GMP: none
CHPL_HWLOC: hwloc
CHPL_REGEXP: re2
CHPL_WIDE_POINTERS: struct
CHPL_AUX_FILESYS: none
```
- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:
```
>  gcc --version
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.5) 5.4.0 20160609
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

### PR DESCRIPTION
**Pull request by nimitbhardwaj**
When write a record using `writef("%jt", obj);`, then the object is not written correctly if enum is
involved in the record.

closes #8252 

- [x] full local testing

### CODE REVIEW

**Comment by nimitbhardwaj** [comment]
I haven't added tests for it yet, I also wanted to know, if the changes done don't affect other stuff.

I edited the read and write routines when reading or writing of an enum was involved, this is done in case of JSON, so I first edited the write, before it was writing the literal, replacing it to print string, which I think its perfect here and it doesn't produce any bug for now, but as I edited it, the JSONs are now {"a": "Type1"}, for Chapel, in constrast to {"a": Type1}, before.
So I had to edit the reading functions too, so I only thought of scanning string with form "<string>" so perhaps serving purpose here.

I need some feedback and verification.

**Comment by nimitbhardwaj** [comment]
```chapel
enum EnumType1 { Type1, Type2, Type3 };
record RecType {
  const s : string;
  const e : EnumType1;
}


var r = new RecType("hello", EnumType1.Type1);

var mem = openmem();
var writer = mem.writer().write('{"s":"ninja", "e":"Type3"}');
var reader = mem.reader();

var f = new RecType();
reader.readf("%jt", f);

writeln(f);
writef("%jt\n", f);

writef( "%jt\n", new RecType( "example_string", EnumType1.Type2)  );
writeln(r);


```

I used this example for me, this seems work, but for read("%t"), its not working correct, so I need some suggestions how to make it, I think we can make a new function, or make it special for JSON only.

**Comment by mppf** [comment]
I believe this will add quotes for non-json writes.

You'll have to check that the format is JSON, e.g. this code from IO.chpl
``` chapel
    var st = writer.styleElement(QIO_STYLE_ELEMENT_AGGREGATE);
    var iolit:ioLiteral;
    if st == QIO_AGGREGATE_FORMAT_JSON {
      iolit = new ioLiteral("null");
    } else {
      iolit = new ioLiteral("nil");
    }

```

**Comment by nimitbhardwaj** [comment]
@mppf , thanks, I was thinking of it, today i picked it again, I was on this conclusion, your answer proved this, thanks, it will be done by tomorrow, its night here now, I work on it tomorrow, thanks.

**Comment by nimitbhardwaj** [comment]
But one more thing i feared of, can this change affect other tests too, like previous tests based on jsons, perhaps, but not sure, and running each and every test is lengthy on my PC, its slow, compiling on my pc takes 5-6 sec, so what shall I do for this.

**Comment by mppf** [comment]
@nimitbhardwaj - I recommend you check test/io/ferguson and get those passing as a starting point.  Later on, I can run full testing for this PR on your behalf.

**Comment by nimitbhardwaj** [comment]
I make the commit, this time its working, I tested it on my personal tests, and its not violating any test in tests/io/ferguson, I just have to make the test for this change, I want to know where I put this test,

will tests/io/ferguson/json.chpl be fine??

**Comment by mppf** [comment]
@nimitbhardwaj - it's fine with me for you to add your test to test/io/ferguson/ but it's probably worth naming it something like json-enum.chpl.

File: modules/standard/IO.chpl [hunk 1]
```diff
@@ -2984,6 +2990,17 @@ private inline proc _read_one_internal(_channel_internal:qio_channel_ptr_t,
...
+
+  var reader = new channel(writing=false, iokind.dynamic, locking=false,
+                           home=here,
+                           _channel_internal=_channel_internal,
+                           _readWriteThisFromLocale=loc);
+  var inJson = false;
+  var st = reader.styleElement(QIO_STYLE_ELEMENT_AGGREGATE);
+
+  if st == QIO_AGGREGATE_FORMAT_JSON then inJson = true;
+  reader._channel_internal = QIO_CHANNEL_PTR_NULL;
```
Comment (mppf) [inline modules/standard/IO.chpl]: Two things:

1. I can't understand why we'd create a `new channel` at this point. It seems to me we just need to check the styleElement of the current channel. When I pointed you to the example earlier, it didn't occur to me that we wouldn't have the `record channel` available. We can call the function to get the style element on the _channel_internal, with `var st = qio_channel_style_element(_channel_internal, QIO_STYLE_ELEMENT_AGGREGATE);`

2. I don't see a good reason to get the style element here in _read_one_internal/_write_one_internal and pass it to _read_text_internal/_write_text_internal. I think it should just make the query when handling an enum in  _read_text_internal/_write_text_internal.

Comment (nimitbhardwaj) [reply modules/standard/IO.chpl]: OK, I didn't know if there is such a function, earlier in code, as you mentioned before, there was a writer writer created, in some overload of _write_one_internal, but if there is such a function exists then its much better, thanks, its better to just get the st without creating another object.

**Comment by nimitbhardwaj** [comment]
I made the test there, it is correct in my sense, please review it, if I made this correct

**Comment by nimitbhardwaj** [comment]
Is this all correct, or any changes here ?

File: modules/standard/IO.chpl [hunk 2]
```diff
@@ -2820,8 +2824,12 @@ private proc _write_text_internal(_channel_internal:qio_channel_ptr_t, x:?t):sys
     const local_x = x.localize();
     return qio_channel_print_string(false, _channel_internal, local_x.c_str(), local_x.length:ssize_t);
   } else if isEnumType(t) {
+    var st = qio_channel_style_element(_channel_internal, QIO_STYLE_ELEMENT_AGGREGATE);
     var s = x:string;
-    return qio_channel_print_literal(false, _channel_internal, s.c_str(), s.length:ssize_t);
+    if st == QIO_AGGREGATE_FORMAT_JSON then
+      return qio_channel_print_string(false, _channel_internal, s.c_str(), s.length:ssize_t);
+    else
+      return qio_channel_print_literal(false, _channel_internal, s.c_str(), s.length:ssize_t);
```
Comment (mppf) [inline modules/standard/IO.chpl]: I think it'd be nicer if this mirrored the _read_text_internal version:
``` chapel
if st == QIO_AGGREGATE_FORMAT_JSON then s = '"'+s+'"';
```

File: modules/standard/IO.chpl [hunk 3]
```diff
@@ -2984,6 +2992,7 @@ private inline proc _read_one_internal(_channel_internal:qio_channel_ptr_t,
                                        ref x:?t,
                                        loc:locale):syserr where _isIoPrimitiveTypeOrNewline(t) {
   var e:syserr = ENOERR;
+
```
Comment (mppf) [inline modules/standard/IO.chpl]: You can & should remove these spacing changes (since they were from code you added and then removed)

**Comment by mppf** [comment]
Other than my (minor) feedback I think this PR is looking good. I'm going to add some TO-DOs to the PR message that need doing before testing.

**Comment by nimitbhardwaj** [comment]
Fine Sire

**Comment by mppf** [comment]
>  I'm going to add some TO-DOs to the PR message that need doing before testing.

Sorry, I wrote that wrong. I meant - I'm going to add TODOs that need to be done before merging, including testing. Let me know when you think the PR is ready and I'll run testing.

**Comment by nimitbhardwaj** [comment]
I did it, though I could have done it yesterday, but yesterday I had to prepare an important class assignment

**Pull request merged**